docs/src/main/asciidoc/spring-stream.adoc

```diff
@@ -76,6 +76,69 @@ spring.cloud.stream.gcp.pubsub.bindings.events.consumer.auto-create-resources=tr
 spring.cloud.stream.bindings.events.group=consumerGroup1
 ----
 
+=== Binding with Functions
+
+Since version 3.0, Spring Cloud Stream supports a functional programming model natively.
+This means that the only requirement for turning your application into a sink is presence of a `java.util.function.Consumer` bean in the application context.
+
+```
+@Bean
+public Consumer<UserMessage> logUserMessage() {
+  return userMessage -> {
+    // process message
+  }
+};
+```
+
+A source application is one where a `Supplier` bean is present.
+It can return an object, in which case Spring Cloud Stream will invoke the supplier repeatedly.
+Alternatively, the function can return a reactive stream, which will be used as is.
+
+```
+@Bean
+Supplier<Flux<UserMessage>> generateUserMessages() {
+  return () -> /* flux creation logic */;
+}
+```
+
+A processor application works similarly to a source application, except it is triggered by presence of a `Function` bean.
+
+WARNING: When using the `spring-cloud-gcp-pubsub-stream-binder` dependency, configuring `spring.cloud.function.definition` property to indicate function bean is *required*, even though Spring Cloud Stream has ability to autodiscover functional beans.
```

Comment:
This seems like a major headache and defeats the promised simplicity of auto-discovering the function.
How bad would it be to hide the `pubSubReactiveScheduler` bean from the application context by creating it directly inside the `pubSubReactiveFactory` bean definition?

```suggestion
WARNING: Even though Spring Cloud Stream is able to autodiscover functional beans, when using the `spring-cloud-gcp-pubsub-stream-binder` dependency, configuring the `spring.cloud.function.definition` property to indicate the function bean is *required*.
```

Reply:
Need to figure out the lifecycle of reactor schedulers; the reason for having schedulers as beans is to have Spring close them on application shutdown.

I created #1997 to track. 
Once this sample is in, I'll also create an issue in SCS repo, pointing to it as an example of the problem.